hive/hive-export-import-metastore.md

```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
-    done;
-done
-~~~
-
-This will generate a file named `alltables.hql`.
-
-3) Copy the file `alltables.hql` to the new HDInsight cluster and run the following command:
-
-~~~
-beeline -u "jdbc:hive2://localhost:10001/;transportMode=http" -f alltables.hql
-~~~
-
-Note: This assumes that data paths on new cluster are same as on old. If not, you can manually edit the generated  
-`alltables.hql`  file to reflect any changes. *For ACID tables, a new copy of the data will be created*
-
-Note: This script also assumes that once the script is complete, the old cluster will **not** be used any longer 
-
-#### Further Reading:
-
-1) [Connect to HDInsight Cluster using SSH](https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-hadoop-linux-use-ssh-unix)
+Need to export Hive metastore and import it on another HDInsight cluster.
+
+* See [apache-hive-migrate-workloads](https://docs.microsoft.com/en-us/azure/hdinsight/interactive-query/apache-hive-migrate-workloads) for migration of an external metastore, or for upgrading Hive from HDInsight 3.6 to 4.0.
+
+* This current article gives instructions on using a script to export/import contents of an internal Hive metastore.
+
+* Import works only if the destination cluster shares the same Storage Account as the source cluster.
+
+> [!NOTE]
+>
+> * The script supports copying of tables and partitions. For HDInsight 4.0, it additionally covers constraints, views, and materialized views. Other metadata objects, like UDFs, must be copied manually.
+>
+> * All managed tables will become transactional if the output HDInsight version is 4.0. Optionally, make the table non-transactional by exporting the data to an external table with the property 'external.table.purge'='true'. For example,
+>
+>    ```SQL
+>    create table tablename_backup like tablename;
+>    insert overwrite table tablename_backup select * from tablename;
+>    create external table tablename_tmp like tablename;
+>    insert overwrite table tablename_tmp select * from tablename;
+>    alter table tablename_tmp set tblproperties('external.table.purge'='true');
+>    drop table tablename;
+>    alter table tablename_tmp rename to tablename;
+>
+> * This procedure preserves non-ACID table locations. You can manually edit the DDL in
+`alltables.hql`, generated from the script, to reflect any location changes. *For ACID tables, a new copy of the data will be created*.
+>
+> * The procedure assumes that after completion, the old cluster will **not** be used any longer.
+
+#### Resolution Steps:
+
+If migrating from an external metastore, follow steps in [apache-hive-migrate-workloads](https://docs.microsoft.com/en-us/azure/hdinsight/interactive-query/apache-hive-migrate-workloads). Otherwise, follow steps below.
+
+1) [Connect to the HDInsight Cluster using SSH](https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-hadoop-linux-use-ssh-unix), to the primary headnode.
+
+1) Download the export script from an ssh session to the cluster:
```

Comment:
Lets make sure that customers only have to run wget/curl exactly once. We should have customers download and run an overall script that will download everything else on their behalf, including helper scripts or other dependencies.

Reply:
In fact in the bigger picture, it would be ideal for the customer to run some pattern like this:
1) wget the script
2) launch exaclty 1 command

3) copy alltables.hql (can we rename this to something more explanatory? Perhaps we should actually have a wrapper script around alltables.hql that does all the setup automatically) to import-destination cluster
4) run alltables.hql or equivalent command exactly once

Reply:
Fewer steps is a noble vision. Will attach a new wrapper script to blob path to do it all. Will have a separate "Prerequisites" section for setting up the cluster and setting Hive configuration.